lib/Echidna/UI.hs

```diff
@@ -78,25 +80,46 @@ ui :: ( MonadCatch m, MonadRandom m, MonadReader x m, MonadUnliftIO m
   let d' = fromMaybe defaultDict d
   let getSeed = view $ hasLens . to seed . non (d' ^. defSeed)
-  bc <- liftIO $ newBChan 100
   ref <- liftIO $ newIORef defaultCampaign
-  dash <- liftM2 (&&) isTerminal $ view (hasLens . dashboard)
-  timeoutSeconds <- (* 1000000) . fromMaybe (-1) <$> view (hasLens . maxTime)
   let updateRef = use hasLens >>= liftIO . atomicWriteIORef ref
-  let updateUI e = when dash $ readIORef ref >>= writeBChan bc . e
-  waitForMe <- liftIO newEmptyMVar
-  ticker <- liftIO $ forkIO -- update UI every 100ms, instant exit when no UI
-    (when dash $ forever $ threadDelay 100000 >> updateUI CampaignUpdated)
-  _ <- forkFinally -- run worker
-    (void $ timeout timeoutSeconds (campaign updateRef v w ts d txs) >>= \case
-       Nothing -> liftIO $ updateUI CampaignTimedout
-       Just _ -> liftIO $ updateUI CampaignUpdated
-    )
-    (const $ liftIO $ killThread ticker >> putMVar waitForMe ())
-  app <- customMain (mkVty defaultConfig) (Just bc) <$> monitor
-  liftIO $ if dash
-             then void $ app (defaultCampaign, Uninitialized)
-             else takeMVar waitForMe
-  final <- liftIO $ readIORef ref
-  liftIO . putStrLn =<< view (hasLens . finished) <*> pure final <*> getSeed
-  return final
+  timeoutSeconds <- (* 1000000) . fromMaybe (-1) <$> view (hasLens . maxTime)
+  terminalPresent <- isTerminal
+  effectiveMode <- view (hasLens . operationMode) <&> \case
+    Interactive | not terminalPresent -> NonInteractive Text
+    other -> other
+  case effectiveMode of
+    Interactive -> do
+      bc <- liftIO $ newBChan 100
+      let updateUI e = readIORef ref >>= writeBChan bc . e
+      ticker <- liftIO $ forkIO $
+        forever $ threadDelay 100000 >> updateUI CampaignUpdated
+      _ <- forkFinally -- run worker
+        (void $ timeout timeoutSeconds (campaign updateRef v w ts d txs) >>= \case
+          Nothing -> liftIO $ updateUI CampaignTimedout
+          Just _ -> liftIO $ updateUI CampaignUpdated
+        )
+        (const $ liftIO $ killThread ticker)
+      app <- customMain (mkVty defaultConfig) (Just bc) <$> monitor
+      liftIO $ void $ app (defaultCampaign, Uninitialized)
+      -- TODO: this is temporary, show seed in the UI
+      liftIO . putStrLn =<< ("Seed: " ++) . show <$> getSeed
+      liftIO $ readIORef ref
+
+    NonInteractive outputFormat -> do
+      result <- timeout timeoutSeconds (campaign updateRef v w ts d txs)
+      (final, timedout) <- case result of
+        Nothing -> do
+          final <- liftIO $ readIORef ref
+          pure (final, True)
+        Just final ->
+          pure (final, False)
+      case outputFormat of
+        JSON ->
+          liftIO . BS.putStr =<< Echidna.Output.JSON.encodeCampaign final <$> getSeed
+        Text -> do
+          liftIO . putStrLn =<< ppCampaign final
+          liftIO . putStrLn =<< ("Seed: " ++) . show <$> getSeed
+          when timedout $ liftIO $ putStrLn "TIMEOUT!"
+        None ->
+          pure ()
+      pure final
```

Comment:
is the return code correctly preserved when echidna timeouts, right?

Reply:
right, nothing changed here, code is determined in main function based on campaign state